### PR TITLE
[DeviceSanitizer] Support GPU DG2 & GEN Device

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/13450

- Check GPU type by the value of device USM pointer (see "USM Allocation Range" in asan_shadow_setup.cpp), so that we can support more GPU devices automatically
- Added Shadow memory allocation and MemToShadow for DG2

### CODE REVIEW

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, can you help to review this PR?
Thanks very much.

**Comment by AllanZyne** [comment]
Hi @wenju-he @zhaomaosu @yingcong-wu, can you help to review this PR?
Thanks very much!

**Review by wenju-he — APPROVED** [review]
LGTM

**Review by zhaomaosu — APPROVED** [review]
lgtm

File: source/loader/layers/sanitizer/asan_shadow_setup.cpp [hunk 1]
```diff
@@ -121,6 +121,60 @@ ur_result_t DestroyShadowMemory() {
 
 } // namespace pvc
 
+namespace dg2 {
+///
+/// USM Allocation Range
+///   Host/Shared USM : 0x0000_0000_0000_0000 ~ 0x0000_7fff_ffff_ffff
+///   Device      USM : 0xffff_8000_0000_0000 ~ 0xffff_ffff_ffff_ffff
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow_setup.cpp:128]: @wenju-he, this is the USM allocation range of DG2, I afraid there's no document. I confirmed this summary from level zero team.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 2]
```diff
@@ -28,6 +28,15 @@ uptr MemToShadow_CPU(uptr USM_SHADOW_BASE, uptr UPtr) {
     return USM_SHADOW_BASE + (UPtr >> 3);
 }
 
+uptr MemToShadow_DG2(uptr USM_SHADOW_BASE, uptr UPtr) {
+    if (UPtr & 0xFFFF000000000000ULL) { // Device USM
```
[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:32]: The reply is below.

[the remark this answers is not in the document]
Comment (wenju-he) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:32]: I see, thank you @AllanZyne

[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:32]: no problem

File: source/loader/layers/sanitizer/ur_sanitizer_utils.cpp [hunk 3]
```diff
@@ -72,6 +72,20 @@ ur_device_handle_t GetDevice(ur_queue_handle_t Queue) {
     return Device;
 }
 
+std::string GetDeviceName(ur_device_handle_t Device) {
+    size_t Size = 0;
+    [[maybe_unused]] auto Res = context.urDdiTable.Device.pfnGetInfo(
+        Device, UR_DEVICE_INFO_NAME, 0, nullptr, &Size);
+    assert(Res == UR_RESULT_SUCCESS);
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/ur_sanitizer_utils.cpp:79]: NIT: Adding more information with the assert like above implementation?

**Review by yingcong-wu — APPROVED** [review]
overall lgtm.

**Review by yingcong-wu — APPROVED** [review]

**Review by pbalcer — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_shadow_setup.cpp [hunk 4]
```diff
@@ -121,6 +121,60 @@ ur_result_t DestroyShadowMemory() {
...
+///
+/// SHADOW MEMORY MAPPING
+///   Host/Shared USM : 0x0              ~ 0x0fff_ffff_ffff
+///   ?               : 0x1000_0000_0000 ~ 0x1fff_ffff_ffff
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow_setup.cpp:132]:
```suggestion
///   Host/Shared USM : 0x0              ~ 0x0fff_ffff_ffff
///   Device USM              : 0x1000_0000_0000 ~ 0x1fff_ffff_ffff
```

File: source/loader/layers/sanitizer/ur_sanitizer_utils.cpp (hunk 3 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/ur_sanitizer_utils.cpp:79]: Done

**Comment by kbenzie** [comment]
I've removed the *ready to merge* tag for now because the intel/llvm checks are not passing, please add it back again once they are fixed so we can merge safely.

File: source/loader/layers/sanitizer/asan_libdevice.hpp [hunk 5]
```diff
@@ -107,6 +107,10 @@ constexpr auto kSPIR_AsanShadowMemoryGlobalStart =
     "__AsanShadowMemoryGlobalStart";
 constexpr auto kSPIR_AsanShadowMemoryGlobalEnd = "__AsanShadowMemoryGlobalEnd";
 
+constexpr auto kSPIR_AsanShadowMemoryLocalStart =
+    "__AsanShadowMemoryLocalStart";
+constexpr auto kSPIR_AsanShadowMemoryLocalEnd = "__AsanShadowMemoryLocalEnd";
+
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_libdevice.hpp:113]: remove it

**Comment by kbenzie** [comment]
@AllanZyne both this and https://github.com/intel/llvm/pull/13450 have merge conflicts, please resolve those.

**Comment by kbenzie** [comment]
I've removed *ready to merge* until the conflicts are resolved.

**Comment by AllanZyne** [comment]
> I've removed _ready to merge_ until the conflicts are resolved.

Conflicts have been fixed! Thanks!

**Comment by AllanZyne** [comment]
force push to reset the code to merge latest syclos tag

**Comment by kbenzie** [comment]
This [fail](https://github.com/oneapi-src/unified-runtime/actions/runs/10071559807/job/27854472284?pr=1521) is unrelated.

**Pull request merged**